R/step-call-pivot_longer.R

```diff
@@ -0,0 +1,388 @@
+  uses_dot_value <- ".value" %in% names_to
+
+  variable_name <- "variable"
+
+  if (uses_dot_value) {
+    if (!is.null(names_sep)) {
+      .value <- str_separate(measure_vars, into = names_to, sep = names_sep)$.value
+    } else if (!is.null(names_pattern)) {
+      .value <- str_extract(measure_vars, into = names_to, names_pattern)$.value
+    } else {
+      abort("If you use '.value' in `names_to` you must also supply
+            `names_sep' or `names_pattern")
+    }
+
+    v_fct <- factor(.value, levels = unique(.value))
+    measure_vars <- split(measure_vars, v_fct)
+    values_to <- names(measure_vars)
+    names(measure_vars) <- NULL
+
+    if (multiple_names_to) {
+      variable_name <- names_to[!names_to == ".value"]
+    }
+  } else if (multiple_names_to) {
+    if (is.null(names_sep) && is.null(names_pattern)) {
+      abort("If you supply multiple names in `names_to` you must also
+            supply `names_sep` or `names_pattern`")
+    } else if (!is.null(names_sep) && !is.null(names_pattern)) {
+      abort("only one of names_sep or names_pattern should be provided")
+    }
+  } else {
+    variable_name <- names_to
+  }
+
+  args <- list(
+    measure.vars = measure_vars,
+    variable.name = variable_name,
+    value.name = values_to,
+    na.rm = values_drop_na,
+    variable.factor = FALSE
+  )
+
+  # Clean up call args if defaults are used
+  if (variable_name == "variable") {
+    args$variable.name <- NULL
+  }
+
+  if (length(values_to) == 1) {
+    if (values_to == "value") {
+      args$value.name <- NULL
+    }
+  }
+
+  if (is_false(values_drop_na)) {
+    args$na.rm <- NULL
+  }
+
+  id_vars <- sim_vars[!sim_vars %in% unlist(measure_vars)]
+
+  out <- step_call(
+    data,
+    "melt",
+    args = args,
+    vars = c(id_vars, variable_name, values_to)
+  )
+
+  if (!is.null(names_prefix)) {
+    out <- mutate(out, !!variable_name := gsub(paste0("^", names_prefix), "", !!sym(variable_name)))
+  }
+
+  if (multiple_names_to && !uses_dot_value) {
+    if (!is.null(names_sep)) {
+      into_cols <- str_separate(pull(out, !!sym(variable_name)), names_to, sep = names_sep)
+    } else {
+      into_cols <- str_extract(pull(out, !!sym(variable_name)), into = names_to, regex = names_pattern)
+    }
+    out <- mutate(out, !!!into_cols)
+
+    # Need to drop variable_name and move names_to vars to correct position
+    # Recreates relocate logic so only select is necessary, not relocate + select
+    out_vars <- out$vars
+    var_idx <- which(out_vars == variable_name)
+    before_vars <- out_vars[seq_along(out_vars) < var_idx]
+    after_vars <- out_vars[seq_along(out_vars) > var_idx]
+
+    out <- select(out, !!!syms(before_vars), !!!syms(names_to), !!!syms(after_vars))
+  } else if (!multiple_names_to && uses_dot_value) {
+    out <- mutate(out, variable = NULL)
+  }
+
+  out <- step_repair(out, repair = names_repair)
+
+  ## names_ptype & names_transform
+  cast_vars <- intersect(names_to, names(names_ptypes))
```

Comment:
This feels sort of fundamentally un-data.table-y to me. Do you think maybe it's better to just not support these arguments?

Reply:
I think this one is up to your preference. I think you're right - this is functionality that starts to split from what would be considered "normal" data.table. But if you want users of dtplyr to be able to quickly add `lazy_dt()`/`as_tibble()` to their currently existing pipe chains it could be worth keeping.

What do you prefer? Remove or keep them? And if "remove", do you want me to add error messages for them?

Reply:
I think remove — they're not particularly useful arguments so I think few people will use them. You can make an error like "`values_ptype` not supported by dtplyr". 

In general, before embarking on any particularly complex translations (or translations that don't have a good data.table equivalent), I think it's reasonable to determine if people are actually using them. There's little point solving a problem that affects few people when there are missing translations for popular tidyr components.

Reply:
Sounds good. And just to make sure before I change anything - am I dropping both `_ptype` and the `_transform` arguments, or just the `_ptype`?

Reply:
Definitely `_ptype`. I think you could probably skip `_transform` too — I think doing it properly probably means using NSE so you can inline the transformations in to the steps.

Reply:
(Like in `across()` and somewhere else I can't remember)